### PR TITLE
Add File.Append/WriteAllText/Bytes{Async} overloads for span/memory

### LINKED ISSUE #99823: [API Proposal]: Add File.WriteAllBytes taking ReadOnlySpan<byte>
EDITED by @stephentoub on 4/21/2024:
Exact same shape as existing APIs that take `byte[]`/`string`, just taking `ReadOnlySpan<byte/char>` instead for sync and `ReadOnlyMemory<byte/char>` for async.
```diff
public static class File
{
+   public static void WriteAllBytes(string path, ReadOnlySpan<byte> bytes);
+   public static Task WriteAllBytesAsync(string path, ReadOnlyMemory<byte> bytes, CancellationToken cancellationToken = default);

+   public static void WriteAllText(string path, ReadOnlySpan<char> contents);
+   public static void WriteAllText(string path, ReadOnlySpan<char> contents, Encoding encoding);
+   public static Task WriteAllTextAsync(string path, ReadOnlyMemory<byte> contents, CancellationToken cancellationToken = default);
+   public static Task WriteAllTextAsync(string path, ReadOnlyMemory<byte> contents, Encoding encoding, CancellationToken cancellationToken = default);

+   public static void AppendAllBytes(string path, ReadOnlySpan<byte> bytes);
+   public static Task AppendAllBytesAsync(string path, ReadOnlyMemory<byte> bytes, CancellationToken cancellationToken = default);

+   public static void AppendAllText(string path, ReadOnlySpan<char> contents);
+   public static void AppendAllText(string path, ReadOnlySpan<char> contents, Encoding encoding);
+   public static Task AppendAllTextAsync(string path, ReadOnlyMemory<char> contents, CancellationToken cancellationToken = default);
+   public static Task AppendAllTextAsync(string path, ReadOnlyMemory<char> contents, Encoding encoding, CancellationToken cancellationToken = default);
}
```

### Background and motivation

Today, File.ReadAllBytes takes byte[] representing the bytes. It would be good to have an overload that takes ReadOnlySpan<byte> (sync ReadAllBytes) and ReadOnlyMemory<byte> (async ReadAllBytesAsync)

### API Proposal

```csharp
namespace System.IO;

public static class File
{
    public static void WriteAllBytes(string path, ReadOnlySpan<byte> bytes);
    public static void WriteAllBytes(string path, ReadOnlyMemory<byte> bytes);
    public static Task WriteAllBytesAsync(string path, ReadOnlyMemory<byte> bytes, CancellationToken cancellationToken = default);

    public static void AppendAllBytes(string path, ReadOnlySpan<byte> bytes);
    public static void AppendAllBytes(string path, ReadOnlyMemory<byte> bytes);
    public static Task AppendAllBytesAsync(string path, ReadOnlyMemory<byte> bytes, CancellationToken cancellationToken = default);
}
```

### API Usage

```csharp
File.WriteAllBytes(path, "Hello World!"u8);
```

### Risks

_No response_

### PR DESCRIPTION
**Pull request by stephentoub**
Closes https://github.com/dotnet/runtime/issues/99823

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/IO/File.cs [hunk 1]
```diff
@@ -632,13 +632,63 @@ public static string ReadAllText(string path, Encoding encoding)
...
+        /// <exception cref="ArgumentException"><paramref name="path"/> is empty.</exception>
+        /// <exception cref="PathTooLongException">The specified path, file name, or both exceed the system-defined maximum length.</exception>
+        /// <exception cref="DirectoryNotFoundException">The specified path is invalid (for example, it is on an unmapped drive).</exception>
+        /// <exception cref="IOException">An I/O error occurred while opening the file.</exception>
+        /// <exception cref="UnauthorizedAccessException"><paramref name="path"/> specified a file that is read-only.</exception>
+        /// <exception cref="UnauthorizedAccessException"><paramref name="path"/> specified a file that is hidden.</exception>
+        /// <exception cref="UnauthorizedAccessException"><paramref name="path"/> specified a directory.</exception>
+        /// <exception cref="UnauthorizedAccessException">This operation is not supported on the current platform.</exception>
+        /// <exception cref="UnauthorizedAccessException">The caller does not have the required permission.</exception>
```
Comment (Jozkee) [inline src/libraries/System.Private.CoreLib/src/System/IO/File.cs:650]:
```suggestion
        /// <exception cref="UnauthorizedAccessException"><paramref name="path"/> specified a file that is read-only.
        /// -or- <paramref name="path"/> specified a file that is hidden.
        /// -or- <paramref name="path"/> specified a directory.
        /// -or- This operation is not supported on the current platform.
        /// -or- The caller does not have the required permission.</exception>
```
I haven't seen triple-slash docs declaring the same exception multiple times. I know we've done it [this way](https://github.com/dotnet/runtime/blob/d7ae8c61f0937df4dcffe88ba96eae65fa523a78/src/libraries/System.Private.CoreLib/src/System/IO/File.cs#L162-L164) in several places. cc @carlossanlop

**Review by jozkee — APPROVED** [review]

**Comment by Jozkee** [comment]
CI error seems related:
```
Process terminated. Assertion failed.
   at System.IO.RandomAccess.QueueAsyncWriteFile(SafeFileHandle handle, ReadOnlyMemory`1 buffer, Int64 fileOffset, CancellationToken cancellationToken, OSFileStreamStrategy strategy) in /_/src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs:line 375
   at System.IO.File.<WriteAllBytesAsync>g__Core|103_0(String path, ReadOnlyMemory`1 bytes, CancellationToken cancellationToken) in /_/src/libraries/System.Private.CoreLib/src/System/IO/File.cs:line 1281
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilderCore.cs:line 38
   at System.IO.File.WriteAllBytesAsync(String path, ReadOnlyMemory`1 bytes, CancellationToken cancellationToken) in /_/src/libraries/System.Private.CoreLib/src/System/IO/File.cs:line 1274
   at System.IO.Tests.File_ReadWriteAllBytesAsync.EmptyContentCreatesFileAsync() in /_/src/libraries/System.Runtime/tests/System.IO.FileSystem.Tests/File/ReadWriteAllBytesAsync.cs:line 45
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s) in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs:line 290
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/ExecutionContext.cs:line 179
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread) in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs:line 354
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext() in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs:line 350
   at Xunit.Sdk.AsyncTestSyncContext.<>c__DisplayClass7_0.<Post>b__0() in /_/src/xunit.execution/Sdk/AsyncTestSyncContext.cs:line 58
   at Xunit.Sdk.XunitWorkerThread.<>c.<QueueUserWorkItem>b__5_0(Object _) in /_/src/common/XunitWorkerThread.cs:line 37
   at System.Threading.Tasks.Task.InnerInvoke() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:line 2402
   at System.Threading.Tasks.Task.<>c.<.cctor>b__292_0(Object obj) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:line 2382
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/ExecutionContext.cs:line 179
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:line 2342
   at System.Threading.Tasks.ThreadPoolTaskScheduler.<>c.<.cctor>b__10_0(Object s) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ThreadPoolTaskScheduler.cs:line 35
   at System.Threading.Thread.StartCallback() in /_/src/coreclr/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs:line 100
----- end Thu 06/13/2024 22:21:34.45 ----- exit code -2146232797 ----------------------------------------------------------
```

**Comment by stephentoub** [comment]
> CI error seems related

Yes, it looks like it's an existing gap with the existing RandomAccess APIs, when being passed an empty `Memory<T>`. It ends up pinning and getting back a null pointer, which the assert doesn't like. I'll fix it as part of this.

File: src/libraries/System.Private.CoreLib/src/System/IO/File.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/IO/File.cs:650]: I believe this is an acceptable format, and backend tooling generates the "-or-"s.

**Pull request merged**